### PR TITLE
Make op_system callable JAX-native for diffrax/NUTS workflows

### LINKED ISSUE #74: Make op_system callable JAX-native for diffrax/NUTS workflows
## Summary
`op_system` currently compiles a callable RHS (`eval_fn`) that is NumPy-coercing at multiple points. This makes the resulting callable non-JAX-traceable, which blocks true diffrax-native + gradient-based inference workflows (e.g. NUTS/HMC through the ODE solve).

## Problem
The current callable path concretizes values to NumPy/Python scalars in both compiler and system wrapper layers:
- `src/op_system/compile.py`
  - env binding uses `np` + `np.float64(...)`
  - helper reductions cast via `float(...)` / `np.float64(...)`
  - equation evaluation builds outputs with `np.empty(..., dtype=np.float64)` + scalar assignment
  - `eval_fn` coercions use `np.asarray(..., dtype=np.float64)`
- `flepimop2-op_system/src/flepimop2/system/op_system/__init__.py`
  - `_stepper` coercions force NumPy for `time`, `state`, and return value

These coercions break tracer propagation and prevent end-to-end JAX differentiation.

## Proposed Solution
Introduce a backend-aware compile/eval path that preserves array types:

1. Add backend namespace parameter to compile path
- Extend `compile_rhs` / `_make_eval_fn` with a backend namespace arg (e.g., `xp`) defaulting to NumPy.
- Bind expression env so user expressions using `np.*` still work by mapping `np -> xp`.

2. Remove scalar concretization in compiler runtime
- In alias/equation evaluation, avoid `float(...)` and `np.float64(...)` casts.
- Keep values as backend array/scalar types.
- Replace reductions with backend ops (`xp.sum(xp.stack(...))`) instead of Python `sum(float(...))`.

3. Make equation output backend-preserving
- Replace `np.empty(..., dtype=np.float64)` + indexed assignment with list collection + `xp.asarray(...)`.

4. Relax `eval_fn` coercions
- Avoid unconditional `np.asarray(..., dtype=np.float64)` and `np.float64(t)` in JAX mode.
- Keep validation but do not force conversion that concretizes tracers.

5. Update flepimop2-op_system `_stepper`
- Keep shape checks.
- Remove forced NumPy casts around input/output in JAX mode.
- Return compiled callable output without downcasting to NumPy.

## Scope / Compatibility
- Preserve existing behavior by default (`xp=np`) so current NumPy users are unaffected.
- Add optional JAX-native path (e.g., `xp=jax.numpy`) for diffrax-native integration.

## Acceptance Criteria
- Existing NumPy workflows remain unchanged and pass current tests.
- A JAX-enabled compiled callable can be evaluated under `jax.jit` without concretization errors.
- diffrax integration can consume the callable without `jax.pure_callback`.
- Gradient-based inference (NUTS/HMC) can trace through RHS evaluation for a representative model.

## Suggested Follow-up
- Add focused tests for both backends:
  - NumPy parity tests (old vs new behavior)
  - JAX tracer tests (`jit`, `grad` where applicable)
  - Minimal diffrax solve smoke test using compiled callable

### PR DESCRIPTION
**Pull request by jc-macdonald**
Closes #74

## Summary

Adds a backend-aware compile path so `compile_rhs` (and `compile_spec`) can return RHS callables that are fully traceable by JAX — no forced NumPy concretization.

### Core (`op_system`)
- `compile_rhs(..., xp=np)` / `compile_spec(..., xp=None)` — new optional `xp` parameter; defaults to `numpy` (no behaviour change for existing callers)
- `_is_numpy_backend(xp)` guard routes array ops through `xp.sum(xp.stack(...))` / `xp.asarray()` for JAX compatibility
- `_BackendNamespace` / `_Indexable` internal protocols replace `Any` in all public callable signatures
- `eval_fn` skips forced `float64` cast in JAX mode (avoids `x64` config requirement)

### Adapter (`flepimop2-op_system`)
- New `backend: Literal["numpy", "jax"]` field on the stepper model
- `_get_backend_namespace()` lazy-imports `jax.numpy` via `importlib` (optional dep, no import-time cost)
- Stepper coerces arrays via NumPy on numpy path; passes through on JAX path

### Tests
- `test_compile_rhs_with_jax_backend_is_jittable` — verifies `jax.jit` wrapping round-trips
- `test_compile_rhs_with_diffrax_solve` — ODE solve through diffrax `ODETerm`
- `test_compile_rhs_traces_through_blackjax_nuts` — NUTS gradient tracing smoke test
- `test_jax_backend_stepper_jittable` (adapter) — stepper survives `jax.jit`
- All new tests use `pytest.importorskip` and are skipped when JAX/diffrax/blackjax are absent

### Packaging
- `jax` and `jax-inference` optional dependency groups added to `pyproject.toml`
- README updated with install instructions and usage examples

### CODE REVIEW

File: flepimop2-op_system/src/flepimop2/system/op_system/__init__.py [hunk 1]
```diff
@@ -65,6 +66,7 @@ class _AxesMeta(NamedTuple):
 class OpSystemSystem(ModuleModel, SystemABC):  # noqa: D101
     module: Literal["flepimop2.system.op_system"] = "flepimop2.system.op_system"
     state_change: StateChangeEnum = StateChangeEnum.FLOW
+    backend: Literal["numpy", "jax"] = "numpy"
```
Comment (pearsonca) [inline flepimop2-op_system/src/flepimop2/system/op_system/__init__.py]: 1. wary of overloading the term "backend" - alternatives?
2. relative to how we're aiming to share non-standard info from custom modules: is this something an engine might need to access? if so, should we be using the options infrastructure?

Comment (pearsonca) [reply flepimop2-op_system/src/flepimop2/system/op_system/__init__.py]: (i *think* the answer to 2 may be no, but perhaps not?)

File: src/op_system/__init__.py [hunk 2]
```diff
@@ -50,20 +50,23 @@
 # -----------------------------------------------------------------------------
 
 
-def compile_spec(spec: dict[str, object]) -> CompiledRhs:  # noqa: RUF067
+def compile_spec(spec: dict[str, object], *, xp: object | None = None) -> CompiledRhs:  # noqa: RUF067
```
Comment (pearsonca) [inline src/op_system/__init__.py]: I think we want to have defaults specified at one entry point. Is this or compile_rhs the right entry point?

File: src/op_system/compile.py [hunk 3]
```diff
@@ -379,88 +404,108 @@ def _resolve_aliases(
...
+    return cast("Float64Array", cast("_BackendNamespace", xp).asarray(out_vals))
 
 
-def _make_eval_fn(
+def _make_eval_fn(  # noqa: C901
     *,
     state_names: tuple[str, ...],
     aliases: Mapping[str, str],
     equations: tuple[str, ...],
+    xp: object,
```
Comment (pearsonca) [inline src/op_system/compile.py:452]: can this be passing a _BackendNamespace instead of object?

**Review by pearsonca — COMMENTED** [review]
Big picture items:

 - need to fall into the "standard" approach to configuration
 - seems like we should be able to be more specific about types relative to the different backends

File: flepimop2-op_system/src/flepimop2/system/op_system/__init__.py (hunk 1 above)
Comment (pearsonca) [reply flepimop2-op_system/src/flepimop2/system/op_system/__init__.py]: per discussion on zoom, the answer to 2 is yes, so we need to fall into the "options" approach

File: src/op_system/compile.py (hunk 3 above)
Comment (pearsonca) [reply src/op_system/compile.py:452]: (this is getting into the typing consideration - but basically look for "object" anywhere in the code)

**Comment by jc-macdonald** [comment]
Thanks for the thorough review — here's a summary of what changed in the latest push (b24138a) and what's tracked for follow-up.

**Standard configuration approach (options infrastructure)**

Removed the model-level `backend` field from `OpSystemSystem`. Array namespace selection is now driven entirely through `options` using the key `array_backend`. The adapter reads and validates the value from `options` at init time, then stores the normalized value back into `options` so engine code can access it consistently via `sys.option("array_backend")`. Defaults to `"numpy"` when omitted.

Relevant code: `_resolve_array_backend` / `_get_array_namespace` in `flepimop2-op_system/src/flepimop2/system/op_system/__init__.py` (lines 176–194), `options` dict construction (lines 95–109).
Test coverage: `test_option_array_backend_defaults_to_numpy`, `test_option_array_backend_invalid_value_raises` in `flepimop2-op_system/tests/test_system.py`.

**Default policy ownership**

`compile_spec` is now the single owner of the default backend namespace (`xp=np`). `compile_rhs` no longer carries a default — callers must pass `xp` explicitly. This ensures default policy is not split across entry points.

Relevant code: `src/op_system/__init__.py` line 55, `src/op_system/compile.py` line 518.
Test coverage: `test_compile_spec_owns_default_backend_policy`, `test_compile_rhs_requires_explicit_backend_namespace` in `tests/op_system/test_op_system_compile.py`.

**Naming**

The term `backend` was dropped at the model-field level. I used `array_backend` in options to be more specific and reduce ambiguity with other uses of "backend" in the broader stack. Further terminology work if needed is tracked in #79.

**Remaining typing work**

The broader `object` annotation cleanup across backend-sensitive internals in `compile.py` is deferred — tracked in #78. Kept this PR scoped to the config pattern and entry-point ownership to land those changes cleanly first.

File: flepimop2-op_system/src/flepimop2/system/op_system/__init__.py [hunk 4]
```diff
@@ -77,7 +78,10 @@ def model_post_init(self, context: Any) -> None:  # noqa: ANN401
         del context
 
         spec_obj = self.spec
-        compiled = compile_spec(spec_obj)
+        existing_options = dict(getattr(self, "options", {}) or {})
```
Comment (pearsonca) [inline flepimop2-op_system/src/flepimop2/system/op_system/__init__.py]: ModuleModel's have an "option" method: https://github.com/ACCIDDA/flepimop2/blob/8a61292f479903bff5a754e98381631b3579c414/src/flepimop2/module.py#L44

Comment (pearsonca) [reply flepimop2-op_system/src/flepimop2/system/op_system/__init__.py]: but i suppose you're using the works in a later step - might be reasonable to keep getting this way.

might also be preferrable to initialize to an empty dict; @TimothyWillard recall why we went with None instead of empty container?

File: flepimop2-op_system/src/flepimop2/system/op_system/__init__.py [hunk 5]
```diff
@@ -100,14 +105,18 @@ def model_post_init(self, context: Any) -> None:  # noqa: ANN401
             "mixing_kernels": mixing_kernels,
             "operators": operators,
             "operator_axis": operator_axis,
+            "array_backend": array_backend,
         }
```
Comment (pearsonca) [inline flepimop2-op_system/src/flepimop2/system/op_system/__init__.py:122]: @TimothyWillard suggests we might want to add a post-initialization method to add elements to ModuleModel.options?

File: flepimop2-op_system/src/flepimop2/system/op_system/__init__.py [hunk 6]
```diff
@@ -100,14 +105,18 @@ def model_post_init(self, context: Any) -> None:  # noqa: ANN401
...
         def _stepper(
             time: np.float64,
             state: Float64NDArray,
             **kwargs: Any,  # noqa: ANN401
         ) -> Float64NDArray:
-            state_arr = np.asarray(state, dtype=np.float64)
+            if array_backend == "numpy":
+                state_arr = np.asarray(state, dtype=np.float64)
+            else:
+                state_arr = xp.asarray(state)
```
Comment (pearsonca) [inline flepimop2-op_system/src/flepimop2/system/op_system/__init__.py]: could the servicing module be stored as an object property, and then used in that fashion instead of this control fork?

**Review by pearsonca — APPROVED** [review]
Minor additional fixes noted; can address those now or as clean-up issue.

File: flepimop2-op_system/src/flepimop2/system/op_system/__init__.py (hunk 4 above)
Comment (TimothyWillard) [reply flepimop2-op_system/src/flepimop2/system/op_system/__init__.py]: `ModuleModel` do not support options, the options infrastructure is implemented by `ModuleABC`. The reason for not making the default an empty dict is because that would require custom initialization that would require some additional unit testing when supporting both pydantic and non-pydantic modules.

However, accessing the `options` attribute directly is less than ideal, it would be preferable that developers use the `ModuleABC.option` method: https://accidda.github.io/flepimop2/latest/reference/api/module/#flepimop2.module.ModuleABC.option

File: flepimop2-op_system/src/flepimop2/system/op_system/__init__.py (hunk 5 above)
Comment (TimothyWillard) [reply flepimop2-op_system/src/flepimop2/system/op_system/__init__.py:122]: Probably not, I think for pydantic modules users should leverage the hooks available there and for non-pydantic modules to do it directly in the user defined `__init__` method.

File: flepimop2-op_system/src/flepimop2/system/op_system/__init__.py (hunk 4 above)
Comment (pearsonca) [reply flepimop2-op_system/src/flepimop2/system/op_system/__init__.py]: right, ModuleABC (which also arrives here via SystemABC)

File: pyproject.toml [hunk 7]
```diff
@@ -33,6 +33,14 @@ data = [
     "pandas>=2.2",
     "pyarrow>=16.0",
 ]
+jax = [
+    "jax>=0.4",
+]
+jax-inference = [
+    "jax>=0.4",
```
Comment (TimothyWillard) [inline pyproject.toml:40]:
```suggestion
    "op-system[jax]",
```

Then only have to maintain `jax` dep in one place

File: src/op_system/compile.py [hunk 8]
```diff
@@ -34,9 +42,26 @@
...
+class _BackendNamespace(Protocol):
+    def asarray(self, obj: object, dtype: object | None = None) -> object: ...
+
+    def stack(self, arrays: list[object]) -> object: ...
+
+    def sum(self, arr: object) -> object: ...
+
+
+class _Indexable(Protocol):
+    def __getitem__(self, idx: int) -> object: ...
```
Comment (TimothyWillard) [inline src/op_system/compile.py:58]: I think the need for this protocol could be alleviated if instead of converting the numpy array of float types to jax typing objects, see https://docs.jax.dev/en/latest/jax.typing.html. I think maybe https://docs.jax.dev/en/latest/_autosummary/jax.typing.ArrayLike.html#jax.typing.ArrayLike is what you're reaching for? However, not sure if this will create knock on effects with typing.

File: README.md [hunk 9]
```diff
@@ -28,6 +28,36 @@ spec = {
...
+
+Then pass the backend namespace when compiling:
+
+```python
+import jax.numpy as jnp
+from op_system import compile_spec
+
+compiled = compile_spec(spec, xp=jnp)
+dydt = compiled.eval_fn(0.0, jnp.asarray([999.0, 1.0, 0.0]), beta=0.3, gamma=0.1)
+```
```
Comment (TimothyWillard) [inline README.md:57]: I think the `xp=...` syntax is a bit odd to me? I would have thought the way to handle this would be to have a `backend: Literal["jax", "numpy"] = "numpy"` (or whatever you want to name the arg) and then behind the scenes attempt to import jax and if not possible raise an `ImportError` that explains to the user they need to do `pip install "op_system[jax]"`. The behind the scenes helper would be something like `_resolve_xp(backend: Literal["jax", "numpy"]) -> types.ModuleType`, or your `_BackendNamespace`.

**Review by TimothyWillard — APPROVED** [review]
Looks fine, mostly made comments about the typing and one about the user UX. Probably some follow up work required there.

I think the underlying array type that is used by modules is something that should be elevated to a `state_change` like property, something that modules have to expose and determines if their compatible or not. I think this gets even easier with https://github.com/ACCIDDA/flepimop2/pull/209 since the `ParameterValue` could be converted to a generic, and I'm assuming we'll have similarly structured containers that can be generics of the underlying array type.

**Comment by jc-macdonald** [comment]
Thanks both — pushed a follow-up commit to address the remaining actionable review comments: dd4a949.

What this update addresses now:
- Adapter options usage/style cleanup:
  - Uses option-based access path for backend selection (array_backend) and normalizes into options during init.
- Runtime control-flow cleanup in adapter:
  - Backend-specific coercion/runtime behavior is centralized via helper setup instead of repeated branching logic.
- Packaging DRY for JAX extras:
  - Added adapter extras that delegate to op-system extras so JAX dependency declarations aren’t duplicated.
- API UX improvement at compile entrypoint:
  - compile_spec now supports a backend selector path with internal resolution while still allowing explicit xp override.
- Docs/tests updated accordingly:
  - README now documents backend-selector usage.
  - compile/adapter tests updated and expanded for the new behavior.

What remains intentionally tracked:
- Issue #78: typing tightening in compile backend paths (object -> stronger types/protocols).
- Issue #79: broader terminology consistency and backend naming alignment.
